documentation/porting_guidelines.md

```diff
@@ -0,0 +1,400 @@
+to store a key in software, especially not the key that is the root for
+everything else, so this is not something we recommend that you should do.
+
+## 4. Secure Clock
+The Time API in GlobalPlatform Internal Core API specification defines three
+sources of time; system time, TA persistent time and REE time. The REE time
+is by nature considered as an unsecure source of time, but the other two should
+in a fully trustable hardware make use of trustable source of time, i.e., a
+secure clock. Note that from GlobalPlatform point of view it is not required to
+make use of a secure clock, i.e., it is OK to use time from REE, but the level
+of trust should be reflected by the `gpd.tee.systemTime.protectionLevel`
+property and the `gpd.tee.TAPersistentTime.protectionLevel` property (100=REE
+controlled clock, 1000=TEE controlled clock). So the functions that one needs to
+pay attention to are `tee_time_get_sys_time()` and `tee_time_get_ta_time()`. If
+your hardware has a secure clock, then you probably want to change the
+implementation there to instead use the secure clock (and then you'd also need
+to update the property accordingly, i.e.,
+`tee_time_get_sys_time_protection_level()` and the variable `ta_time_prot_lvl`
+in `tee_svc.c`).
+
+## 5. Root and Chain of Trust
+To be able to assure that your device are running the (untampered) binaries you
+intended to run you will need to establish some kind of trust anchor on the
+devices. The most common way of doing that is to put the root public key in some
+read only memory on the device. Quite often SoC's/OEM's stores the hash of the
+public key(s) in [OTP]. When the boot ROM (which indeed needs to be ROM) is
+about to load the first stage bootloader it typically reads the public key from
+the software binary itself, hash the key and compare it to the key in OTP. If
+they are matching, then the boot ROM can be sure that the first stage bootloader
+was indeed signed with the corresponding private key.
+
+In OP-TEE you won't find any code at all related to this and this is a good
+example when it is hard for us to do this in a generic way since device
+manufacturers all tend to do this in their own unique way and they are not very
+keen on sharing their low level boot details and security implementation with
+the rest of the world. This is especially true on ARMv7-A. For ARMv8-A it looks
+bit better, since ARM in ARM Trusted Firmware have implemented and defined how a
+abstract the chain of trust (see [auth-framework.md]). We have successfully
+verified OP-TEE by using the authentication framework from ARM Trusted Firmware
+(see [optee_with_auth_framework.md] for the details).
+
+## 6. Hardware Crypto IP
+By default OP-TEE uses a software crypto library (currently LibTomCrypt) and you
+have the ability to enable Crypto Extensions that was introduced with ARMv8-A
+(if the device is capable of that). To this date, we haven't had any hardware in
```

Comment:
This sentence (`To this date, we haven't had any hardware in our possession where we have had access to the real crypto hardware.`) is redundant with the next sentence that follows (`Some of the devices we have in our hands...`).
I think you could remove the former.